TESTS/block_device/qspif/main.cpp

```diff
@@ -0,0 +1,328 @@
+#include "mbed.h"
+
+using namespace mbed;
+
+#include "greentea-client/test_env.h"
+#include "unity.h"
+#include "utest.h"
+#include "QSPIFBlockDevice.h"
+#include <stdlib.h>
+
+using namespace utest::v1;
+
+#define TEST_BLOCK_COUNT 10
+#define TEST_ERROR_MASK 16
+
+const struct {
+    const char *name;
+    bd_size_t (BlockDevice::*method)() const;
+} ATTRS[] = {
+    {"read size",    &BlockDevice::get_read_size},
+    {"program size", &BlockDevice::get_program_size},
+    {"erase size",   &BlockDevice::get_erase_size},
+    {"total size",   &BlockDevice::size},
+};
+
+static SingletonPtr<PlatformMutex> _mutex;
+
+
+// Mutex is protecting rand() per srand for buffer writing and verification.
+// Mutex is also protecting printouts for clear logs.
+// Mutex is NOT protecting Block Device actions: erase/program/read - which is the purpose of the multithreaded test!
+void basic_erase_program_read_test(
+		QSPIFBlockDevice &blockD,
+		bd_size_t block_size,
+		uint8_t *write_block,
+		uint8_t *read_block,
+		uint8_t *error_mask,
+		unsigned addrwidth)
+{
+	int err = 0;
+    // Find a random block
+     bd_addr_t block = (rand()*block_size) % blockD.size();
+
+     // Use next random number as temporary seed to keep
+     // the address progressing in the pseudorandom sequence
+     unsigned seed = rand();
+
+     // Fill with random sequence
+     _mutex->lock();
+     srand(seed);
+     for (bd_size_t i_ind = 0; i_ind < block_size; i_ind++) {
+         write_block[i_ind] = 0xff & rand();
+     }
+     // Write, sync, and read the block
+     printf("\ntest  %0*llx:%llu...", addrwidth, block, block_size);
+     _mutex->unlock();
+
+     err = blockD.erase(block, block_size);
+     TEST_ASSERT_EQUAL(0, err);
+
+     err = blockD.program(write_block, block, block_size);
+     TEST_ASSERT_EQUAL(0, err);
+
+     _mutex->lock();
+     printf("\nwrite %0*llx:%llu ", addrwidth, block, block_size);
+     for (int i_ind = 0; i_ind < 16; i_ind++) {
+         printf("%02x", write_block[i_ind]);
+     }
+     printf("...\n");
+     _mutex->unlock();
+
+     err = blockD.read(read_block, block, block_size);
+     TEST_ASSERT_EQUAL(0, err);
+
+     _mutex->lock();
+     printf("read  %0*llx:%llu ", addrwidth, block, block_size);
```

Comment:
Assuming these debug prints will be removed later

Reply:
indeed will be removed later on